ksc.1
.TH ksc "1" "Jan 2023" "ksc - Version 1.12" "User Commands"
.SH NAME
ksc \- report symbols used by kernel modules
.SH SYNOPSIS
.B ksc -k \fIFILE\fR ... [OPTIONS]

.B ksc -k \fIFILE\fR ... -s [OPTIONS]

.B ksc -s -p \fIKSC_RESULT\fR [OPTIONS]

.B ksc -S -p \fIKSC_RESULT\fR [OPTIONS]

.SH DESCRIPTION
.B ksc
produces and submits reports of symbols used by kernel object \fIFILE\fRs
(specified by -k). Resulting report is always saved to \fI~/ksc-result.txt\fR.

For more on report generation, see -k.

For more on report submission for stabilization, see -s.

For more on report submission for notification, see -S.

.SH ENVIRONMENT

It is highly recommended that ksc is used on the latest minor version of RHEL,
and to make sure, prior to ksc usage, to update kernel-devel-provided
Module.symvers. This can be accomplished, for example, by issuing:

yum \fBupdate\fR \fBkernel-devel\fR

It is also possible to use the -y (--symvers) option to explicitly state the
location of Module.symvers file to make sure that accurate and up-to-date
kABI data be provided to ksc.

.SH OPTIONS
You may freely mix different option styles.

.TP
\fB\-h\fR, \fB\-\-help\fR
Show help message and exitfor notification.
.TP
\fB\-c\fR \fICONFIG\fR, \fB\-\-\fRconfig=\fICONFIG\fR
Read config options from \fICONFIG\fR file. If not supplied, \fBksc\fR will use
~/ksc.conf. If not found, /etc/ksc.conf will be used instead.
.TP
\fB\-j\fR \fIKSC_RESULT\fR, \fB\-\-\fRjustification-from=\fIKSC_RESULT\fR
Read symbol justifications from \fIKSC_RESULT\fR file and use them in the new
report. It is important to note that placeholder justifications and reference
justifications are ignored by this option and re-generated every time.

This option may be specified multiple times. \fIKSC_RESULT\fR file order is
determined by the order of arguments passed to ksc. If there are two conflicting
non-placeholder/non-reference justifications for a kernel module A.ko symbol S
in multiple report files at the same time, it will be carried over from the
last one specified on the command line.

Symbol justifications can be carried over within the same kernel module only.
Justifications do not propagate through references (i.e., if you justify
usage of S for A.ko, B.ko references justification in A.ko, and you run ksc
without A.ko, justification is not carried over and you will be asked to
specify).
.TP
\fB\-k\fR \fIFILE\fR, \fB\-\-ko\fR=\fIFILE\fR
Collect used symbols in kernel object \fIFILE\fR.
.br
Multiple -k arguments may be supplied, in which case different \fIFILE\fRs are
placed in distinct sections of the \fI~/ksc-result.txt\fR file.

All kernel object \fIFILE\fRs must be build against the same architecture.

If multiple \fIFILE\fRs are supplied, ksc report won't include symbols exported
by any one of the \fIFILE\fRs. This allows for bulk collection of symbols of
a functionally dependent set of kernel modules without having to filter
symbols coming from other modules.
.TP
\fB\-K\fR \fIFILE\fR, \fB\-\-ko-dependency\fR=\fIFILE\fR
Omit any symbols exported by \fIFILE\fR from the report. Use this option if
other modules (which are part of the report) depend on symbols exported by
\fIFILE\fR, but \fIFILE\fR itself is not to be reported at this time.
.TP
\fB\-n\fR \fIRELEASE\fR, \fB\-\-name\fR=\fIRELEASE\fR
RHEL \fIRELEASE\fR against which the bug is to be filed, e.g '6.5'.
.TP
\fB\-C\fR, \fB\-\-centos-stream\fR
Target Centos Stream rather than a specific minor release.
.TP
\fB\-p\fR \fIPREVIOUS\fR, \fB\-\-previous\fR=\fIPREVIOUS\fR
Specified \fIPREVIOUS\fR file is marked for submission to Red Hat Bugzilla.
.br
To submit, you need to use -s or -S.
.TP
\fB\-r\fR \fIRELEASE\fR, \fB\-\-release\fR=\fIRELEASE\fR
RHEL \fIRELEASE\fR used for symbol comparison with stablelists, e.g. '6.5'.
.TP
\fB\-y\fR \fISYMVERS\fR, \fB\-\-symvers\fR=\fISYMVERS\fR
Path to Module.symvers \fISYMVERS\fR file used to check symbol presence in the
the kernel. If not specified, current kernel version is used to determine the
path and Module.symvers of the currently running kernel is used.

You are notified whenever symbol is not present in the Module.symvers file.
.TP
\fB\-s\fR, \fB\-\-submit\fR, \fB\-S\fR, \fB\-\-subscribe\fR
Submit the report to the Red Hat Bugzilla (https://bugzilla.redhat.com).
Credentials used in submission need to be set a config file (see FILES section
below). If API key is not set, ksc will prompt for Bugzilla password.

This option needs to be supplemented with -k or -p.

If used with -k, report gets generated, opened in an editor for you to make
changes, saved in \fI~/ksc-result.txt\fR, and submitted. You may interrupt
submission by pressing ^C.

To submit for stabilization, please use -s. Using -s implies -S.
To submit for notification only, please use -S.
.TP
\fB\-v\fR, \fB\-\-version\fR
Prints ksc version.

.SH FILES

.TP
~/ksc-result.txt
Report produced by ksc.

.TP
~/ksc.conf (/etc/ksc.conf)
Local (global) ksc config file. Global config file is used only when local
config file does not exist. Syntax is as follows:
.IP
.br
[global]
.br
user=user@example.com
.br
partner=none
.br
partnergroup=none
.br
[bugzilla]
.br
enable=1
.br
server=https://bugzilla.redhat.com/xmlrpc.cgi
.br
api_key=API_KEY
.br
[mailing_list]
.br
enable=1
.br
smtp=smtp.example.com
.br
to=kabi-requests@redhat.com

Please replace the values above by your credentials.

Mandatory fields for any submission method: user, partner, partnergroup.
.br
Mandatory fields for Bugzilla submission method: server, api_key.
.br
Mandatory fields for mailing list submission method: to, smtp.

By default, Bugzilla (mailing list) submission method is auto-enabled
(auto-disabled) and must be opted out (in).

Note that server must be a valid XML RPC Bugzilla link, user and to fields
must be valid e-mail addresses.

At least one submission method must be configured to be able to submit
ksc reports.

In case you don't have or don't wish to file the bug under a specific
partner/partnergroup, please set these fields to \fInone\fR (see example
above).

.SH EXAMPLES
.TP
ksc -k a.ko
Process a.ko file using system-provided stablelist and Module.symvers files.
Reports symbols used by a.ko that are not in stablelist, requesting justification
for their use.

.TP
ksc -k a.ko -y /tmp/Module.symvers
Process a.ko file using system-provided stablelist and user-provided
Module.symvers (/tmp/Module.symvers) files. Reports symbols used by a.ko that
are not in stablelist, requesting justification for their use.

.TP
ksc -k a.ko -k b.ko
Process a.ko and b.ko files using system-provided stablelist and Module.symvers
files. Reports symbols used by both a.ko and b.ko that are not in stablelist,
requesting justification for their use. Only one justification per symbol is
required. Any mention of an already justified symbol come with an automatically
generated references.

.TP
ksc -k a.ko -k b.ko -K c.ko -K d.ko
Process a.ko and b.ko files using system-provided stablelist and Module.symvers
files. Reports symbols used by both a.ko and b.ko that are not in stablelist and
not exported by either c.ko or d.ko, requesting justification for their use.
Only one justification per symbol is required. Any mention of an already
justified symbol come with an automatically generated references.

.TP
ksc -k a.ko -k b.ko -j previous-report.txt
Process a.ko and b.ko files using system-provided stablelist and Module.symvers
files. Reports symbols used by both a.ko and b.ko that are not in stablelist,
requesting justification for their use. Only one justification per symbol is
required. Any mention of an already justified symbol come with an automatically
generated references. Justifications already present in previous-report.txt
are carried over to the newly generated one.

.TP
ksc -s -p report-to-submit.txt
Submit a report to bugzilla.



ksc.py
#!/usr/bin/env python3
# Copyright 2012,2018 Red Hat Inc.
# Author: Kushal Das <kdas@redhat.com>

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.  See
# http://www.gnu.org/copyleft/gpl.html for the full text of the
# license.
#
import os
import re
import sys
from optparse import OptionParser, SUPPRESS_HELP
from utils import run, read_list
from utils import read_total_list, get_release_name
from utils import submit_stable, submit_notif
from utils import query_user, query_user_bool

KSCVERSION = "ksc - Version 1.12"

class Ksc(object):

    # RE to detect ksc cli execution
    HEADER_RE                 = re.compile(r"\[command: (?P<cmd>.*)\]")

    # RE to extract KO basename
    SECTION_KO_RE             = re.compile(r'{(?P<ko_file>.*)}')

    # RE to match with KO-body section
    LISTS_RE                  = re.compile(r'.*\[STABLELISTUSAGE\](?P<wl>.*)\[NONSTABLELISTUSAGE\]\s*(?P<gl>.*)', re.S)
    LISTS_RE_DEPRECATED       = re.compile(r'.*\[WHITELISTUSAGE\](?P<wl>.*)\[NONWHITELISTUSAGE\]\s*(?P<gl>.*)', re.S)

    # RE to extract symbol and its justification
    JUSTIFICATION_RE          = re.compile(r'#*\s*\((?P<symbol>.*)\)\s*(?P<justification>[^#]*)')

    # Non-stablelisted symbols justification free-form entries
    JUSTIFICATION_PLACEHOLDER = "ENTER JUSTIFICATION TEXT HERE"
    JUSTIFICATION_REFERENCE   = "JUSTIFICATION STATED UNDER `%s' SECTION"
    JUSTIFICATION_REF_DETECT  = re.compile(r"JUSTIFICATION STATED UNDER `(.*)' SECTION")
    JUSTIFICATION_SEPARATOR   = '#' * 10
    JUSTIFICATION_BODY        = '\n(%s)\n\n%s\n\n'

    # Sections
    # Note: there needn't be old-style declarations here, since this is used
    # for generation only.
    SECTION_STABLELISTS        = "[STABLELISTUSAGE]\n"
    SECTION_CO_STABLELISTS     = "[NONSTABLELISTUSAGE]\n"

    def __init__(self, mock=False):
        """
        Init call
        """
        self.all_symbols_used = {}
        self.nonstable_symbols_used = {}
        self.stable_symbols = {}
        self.defined_symbols = {}
        self.justified_symbols = {}
        self.justifications = {}
        self.matchdata = None
        self.total = None
        self.verbose = False
        self.mock = mock
        self.releasedir = None
        self.symvers = None
        self.arch = None
        self.vermagic = {}
        self.import_ns = {}
        self.releasename = None
        self.centos_stream = False

    def clean(self):
        self.all_symbols_used = {}
        self.nonstable_symbols_used = {}
        self.stable_symbols = {}
        self.defined_symbols = {}
        self.justified_symbols = {}
        self.justifications = {}
        self.matchdata = None
        self.total = None
        self.vermagic = {}
        self.import_ns = {}

    def main(self, mock_options=None):
        """
        Main function for the logic
        """
        filename = os.path.join(os.path.expanduser("~"), "ksc-result.txt")
        # default architecture
        self.arch = "x86_64"

        parser = OptionParser()
        parser.add_option("-c", "--config", dest="config",
                          help="path to configuration file", metavar="CONFIG")
        parser.add_option("-k", "--ko", action="append", dest="ko",
                          help="path to the ko file", metavar="KO")
        parser.add_option("-K", "--ko-dependency", action="append",
                          dest="ko_dependency", help="path to a dependency",
                          metavar="DEPENDENCY")
        parser.add_option("-n", "--name", dest="releasename",
                          help="Red Hat release to file the bug against, "
                               "e.g '6.7'", metavar="RELEASENAME")
        parser.add_option("-C", "--centos-stream", dest="centos",
                          action="store_true", help="Target Centos Stream X")
        parser.add_option("-p", "--previous", dest="previous",
                          help="path to previous resultset to submit as bug")
        parser.add_option("-r", "--release", dest="release",
                          help="RHEL stablelist release to compare against, "
                               "e.g '6.7'", metavar="RELEASE")
        parser.add_option("-y", "--symvers", dest="symvers",
                          help="Path to the Module.symvers file. "
                               "The current kernel path is used if "
                               "not specified.",
                          metavar="SYMVERS")
        parser.add_option("-s", "--submit",
                          action="store_true", dest="submit", default=False,
                          help="Submit to Red Hat Bugzilla")
        parser.add_option("-S", "--subscribe",
                          action="store_true", dest="subscribe", default=False,
                          help="Submit to Red Hat Bugzilla")
        parser.add_option("-v", "--version",
                          action="store_true", dest="VERSION", default=False,
                          help="Prints KSC version number")
        parser.add_option("-j", "--justification-from", action="append",
                          dest="justification_from", metavar="JUSTIFICATION",
                          help="read justification from previous ksc-result")

        if not self.mock:  # pragma: no cover
            (options, args) = parser.parse_args(sys.argv[1:])
        else:  # pragma: no cover
            (options, args) = parser.parse_args(mock_options)

        if options.VERSION:
            print(KSCVERSION)
            sys.exit(0)

        # Create the ksc.conf config path
        if options.config:
            path = os.path.abspath(os.path.expanduser(options.config))
        else:
            path = os.path.expanduser("~/ksc.conf")

        if options.releasename:
            if not self.valid_release_version(options.releasename):
                sys.exit(1)
            self.releasename = options.releasename

        if options.release:
            if not self.valid_release_version(options.release):
                sys.exit(1)

        if options.centos:
            self.centos_stream = options.centos

        if options.releasename and options.release and \
                options.release != options.releasename:
            print("Release and release name do not match.")
            sys.exit(1)

        if options.previous:  # Submit the result of previous run
            filename = os.path.abspath(os.path.expanduser(options.previous))
            if os.path.basename(filename) != 'ksc-result.txt':
                print("Please point to the ksc-result.txt file in -p option.")
                sys.exit(1)

            self.submit(filename, path, options.submit, options.subscribe)
            return

        self.releasedir = 'kabi-current'
        if options.release:
            if not self.valid_release_version(options.release):
                sys.exit(1)

            self.releasedir = 'kabi-rhel' + options.release.replace('.', '')

        if options.symvers:
            self.symvers = options.symvers

        if options.justification_from:
            for file in options.justification_from:
                self.read_justifications(file)

        if options.ko_dependency:
            for kmod_path in options.ko_dependency:
                self.parse_ko(kmod_path, process_stablelists=False)

        if options.ko:
            self.find_arch(options.ko)

            self.read_data(self.arch, self.releasedir, self.symvers)

            for kmod_path in options.ko:
                self.parse_ko(kmod_path, process_stablelists=True)

            self.remove_internal_symbols()

            for kmod_path in options.ko:
                self.print_result(kmod_path)

            self.save_result()

        else:  # pragma: no cover
            print("You need to provide a path to at least one .ko file.")
            sys.exit(1)

        # Now save the result

        if not options.submit and not options.subscribe:
            return

        if not self.mock:  # pragma: no cover
            self.get_justification(filename)
        self.submit(filename, path, options.submit, options.subscribe)

    def read_justifications(self, filepath):
        filepath = os.path.abspath(os.path.expanduser(filepath))

        if not os.path.isfile(filepath):
            print("Filename `%s' does not exist!" % filepath)
            return

        with open(filepath, "r") as fd:

            filename_section = ""

            for file_contents in re.split("({[^}]*})", fd.read()):

                filename_match = self.SECTION_KO_RE.match(file_contents)
                if filename_match:
                    filename_section = filename_match.group('ko_file')

                # Attempt to read new-style justifications, provided they are
                # present. Otherwise assume that old-style justifications were
                # given.
                match = self.LISTS_RE.match(file_contents)
                if not match:
                    match = self.LISTS_RE_DEPRECATED.match(file_contents)
                    if not match:
                        continue

                for symbol, justification in \
                        self.JUSTIFICATION_RE.findall(file_contents):
                    symbol = symbol.strip()
                    justification = justification.strip()

                    if justification == self.JUSTIFICATION_PLACEHOLDER:
                        continue

                    if self.JUSTIFICATION_REF_DETECT.match(justification):
                        continue

                    if filename_section not in self.justifications:
                        self.justifications[filename_section] = {}

                    self.justifications[filename_section][symbol] = \
                            justification

                    if symbol not in self.justified_symbols:
                        self.justified_symbols[symbol] = \
                                os.path.basename(filename_section)

    def valid_release_version(self, release):
        if release == "c9s" or release == "CentOS Stream":
            return True
        rels = release.split(".")
        if len(rels) != 2:
            print("Invalid release: %s" % release)
            return False
        if not rels[0].isdigit() or int(rels[0]) <= 5:
            print("Invalid release: %s" % release)
            return False
        return True

    def submit_get_consent(self):

        """
        Part of the submission process. User gets queried for Red Hat's
        receipt and internal use. Consent is mandatory.
        """

        consent_string = "By using ksc to upload your data to Red Hat, " \
            "you consent to Red Hat's receipt use and analysis of this " \
            "data. Do you agree? [y/N] "

        consent = query_user_bool(consent_string)
        if consent.lower() != 'y':
            print("Cannot proceed without consent. Qutting.")
            sys.exit(1)

    def submit_get_release(self):

        """
        Part of the submission process. User gets queried for release if
        not explicitly provided via argv.
        """

        # Release has been specified as argv, no need to query user at this time
        if self.releasename is not None:
            return

        print("RHEL release not specified with -n flag. Defaulting to your "
            "system's RHEL release.")

        self.releasename = get_release_name()
        use_sys_release  = ""
        if not self.releasename:
            print("Unable to determine system's release name. Please specify.")

        else:
            query = "File bug against RHEL release %s?" % self.releasename
            query += "\n"
            query += "y/N: "

            use_sys_release = query_user_bool(query)

            if not use_sys_release:
                print("Unable to determine user option. Qutting.")
                sys.exit(1)

        # Either system-determine RHEL release is not what user wishes to file
        # against, or ksc couldn't determine the release; query user to specify
        if use_sys_release.lower() == 'n' or not self.releasename:
            release_name = query_user(
                "Please enter valid RHEL release to file bug against: ",
                is_valid=self.valid_release_version
            )

            if not release_name:
                print("Unable to determine a valid RHEL release. Qutting.")
                sys.exit(1)

            else:
                print("Using RHEL %s release." % release_name)

            self.releasename = release_name

    def submit(self, filename, path, submit, subscribe):
        """
        Submits the resultset into Red Hat bugzilla.
        Asks user for Red Hat internal processing consent.
        If not set, determines and asks which RHEL release to use.

        :arg filename: Full path the ksc-result.txt file.
        :arg path: Path to the config file.
        """
        try:
            with open(filename, "r") as fptr:
                line = fptr.readline().strip()
                module_name = self.get_module_name(line)

        except IOError as err:
            print("Unable to read previous result: {}".format(err))
            sys.exit(1)

        if not self.mock:  # Ask for user permission
            self.submit_get_consent()
            self.submit_get_release()

        if not submit and not subscribe:
            print("Please use -s or -S to submit.")
            sys.exit(1)

        if submit:
            return submit_stable(filename, self.arch, mock=self.mock,
                    path=path, releasename=(self.releasename, self.centos_stream),
                    module=module_name)

        submit_notif(filename, self.arch, mock=self.mock, path=path,
                releasename=(self.releasename, self.centos_stream), module=module_name)


    def get_justification(self, filename):
        """
        Get the justification from User
        on non-stablelist symbols

        """
        bold = "\033[1m"
        reset = "\033[0;0m"

        print(bold)
        print('On the next screen, the result log will be opened to allow')
        print('you to provide technical justification on why these symbols')
        print('need to be included in the KABI stablelist.')
        print('Please provide sufficient information in the log, marked with ')
        print('the line below:')

        print(("\n%s\n" % self.JUSTIFICATION_PLACEHOLDER) + reset)
        print(bold + 'Press ENTER for next screen.' + reset)
        try:
            input()
        except EOFError:
            print("Warning. Running in a non-interactive mode.")

        editor = os.getenv('EDITOR')
        if editor:
            os.system(editor + ' ' + filename)
        else:
            os.system('vi ' + filename)
        return True

    def save_result(self):
        """
        Save the result in a text file
        """
        output_filename = os.path.expanduser("~/ksc-result.txt")
        if os.path.isfile(output_filename):

            overwrite_result_query = "ksc-result.txt already exists. " \
                    "Overwrite? [y/N]: "
            overwrite = query_user_bool(overwrite_result_query)

            if overwrite.lower() != 'y':
                print("Unable to get an explicit overwrite acknowledgement. "
                        "Quitting.")
                sys.exit(1)

        if os.path.isfile(output_filename):
            with open(output_filename, 'w+') as f:
                f.truncate()

        try:
            with open(output_filename, "a") as f:
                command = "[command: %s]\n" % " ".join(sys.argv)

                f.write(command)
                for ko_file in self.all_symbols_used:
                    ns = list(filter(lambda x: x, self.import_ns[ko_file]))
                    if ns:
                        ns = "@ns:" + "@ns:".join(ns)
                    else:
                        ns = ""
                    f.write("\n{%s@%s%s}\n\n" % (
                        os.path.basename(ko_file),
                        self.vermagic[ko_file].strip(),
                        ns
                    ))
                    self.write_result(f, ko_file)

            if not self.mock:
                print("A copy of the report is saved in %s" % output_filename)

        except Exception as e:
            print("Error in saving the result file at %s" % output_filename)
            print(e)
            sys.exit(1)

    def print_result(self, kmod_path):
        """
        Print the result (score)
        """

        print("Processing %s" % kmod_path)

        for name in self.nonstable_symbols_used[kmod_path]:
            if name not in self.total:
                print("WARNING: External symbol in %s does not "
                      "exist in current kernel: %s" \
                      % (os.path.basename(kmod_path),name))

        total_len = len(self.all_symbols_used[kmod_path])
        non_stable = len(self.nonstable_symbols_used[kmod_path])
        stable_len = float(len(self.stable_symbols[kmod_path]))

        if total_len == 0:  # pragma: no cover
            print("No kernel symbol usage found in %s." % kmod_path)
            return

        score = (stable_len / total_len) * 100

        if not self.mock:
            print("Checking against architecture %s" % self.arch)
            print("Total symbol usage: %s\t"
                  "Total Non stable list symbol usage: %s"
                  % (total_len, non_stable))
            print("Score: %0.2f%%\n" % score)

    def find_arch(self, kmod_list):
        """
        Finds the architecture of the file in given path
        """
        rset = {'littleendianIntel80386': 'i686',
                'bigendianPowerPC64': 'ppc64',
                'littleendianPowerPC64': 'ppc64le',
                'littleendianAdvancedMicroDevicesX86-64': 'x86_64',
                'bigendianIBMS/390': 's390x',
                'littleendianAArch64': 'aarch64'}
        arch = []
        for kmod_path in kmod_list:
            try:
                data = run("readelf -h %s | grep -e Data -e Machine | awk -F "
                        "':' '{print $2}' | paste -d ' '  - - | awk -F ',' "
                        "'{print $2}' | sed 's/[ \t]*//g'" % kmod_path)
                arch.append(rset[data.strip()])
            except IOError as e:
                print(e, end=' ')
                print(("(Only kernel object files are supported)")
                    if "No such file" not in str(e)
                    else "")
                sys.exit(1)
            except KeyError:
                print("%s: Invalid architecture. (only %s are supported)"
                    % (kmod_path, ', '.join(sorted(rset.values()))))
                sys.exit(1)

        arch = list(set(arch))
        if len(arch) > 1:
            print("Object files for multiple architectures were provided (%s)."
                % ', '.join(sorted(arch)))
            sys.exit(1)

        self.arch = arch[0]

    def write_result(self, f, ko_file):
        """
        Save the result set in the given file
        """
        try:
            ko_basename = os.path.basename(ko_file)

            f.write("[%s]\n" % self.arch)

            # Write stablelisted symbols
            f.write(self.SECTION_STABLELISTS)
            for name in sorted(self.stable_symbols[ko_file]):
                f.write(name + '\n')

            # Write non-stablelisted symbols as well as their justification
            # Justification can be one of:
            #  - free-form entry
            #  - reference to a different kernel module section (if exists)
            #  - justification placeholder later to be specified by hand
            f.write(self.SECTION_CO_STABLELISTS)
            for name in sorted(self.nonstable_symbols_used[ko_file]):

                justification=""
                if name in self.justified_symbols \
                        and ko_basename != self.justified_symbols[name]:
                    justification=self.JUSTIFICATION_REFERENCE % \
                            self.justified_symbols[name]
                elif ko_basename in self.justifications and \
                        name in self.justifications[ko_basename]:
                    justification=self.justifications[ko_basename][name]
                elif "" in self.justifications and \
                        name in self.justifications[ko_basename]:
                    justification=self.justifications[ko_basename][name]
                else:
                    justification=self.JUSTIFICATION_PLACEHOLDER
                    self.justified_symbols[name] = os.path.basename(ko_file)

                f.write(self.JUSTIFICATION_SEPARATOR)
                f.write(self.JUSTIFICATION_BODY % (name, justification))

            if self.nonstable_symbols_used[ko_file]:
                f.write(self.JUSTIFICATION_SEPARATOR)
                f.write('\n')
        except Exception as err:
            print(err)

    def read_data(self, arch, releasedir, symvers):
        """
        Read both data files
        """
        self.matchdata, exists = read_list(arch, releasedir, self.verbose)
        self.total = read_total_list(symvers)
        return exists

    def parse_ko(self, path, process_stablelists=True):
        """
        parse a ko file
        """
        if process_stablelists:
            self.nonstable_symbols_used[path] = set()
            self.all_symbols_used[path] = set()
            self.nonstable_symbols_used[path] = set()
            self.stable_symbols[path] = set()

        self.defined_symbols[path] = set()

        try:
            self.vermagic[path] = run("modinfo -F vermagic '%s'" % path)
        except Exception as e:
            print(e)
            sys.exit(1)

        try:
            self.import_ns[path] = run("modinfo -F import_ns '%s'" % path).split('\n')
        except Exception as e:
            print(e)
            sys.exit(1)

        try:
            out = run("nm '%s'" % path)
        except Exception as e:
            print(e)
            sys.exit(1)

        for line in out.split("\n"):
            data = line.split(" ")
            if len(data) < 2:
                continue
            if "U " in line and process_stablelists:
                self.find_if(path, data[len(data)-1])
            else:
                self.defined_symbols[path].add(data[len(data)-1])

    def remove_internal_symbols(self):
        for i in self.nonstable_symbols_used:
            for j in self.defined_symbols:
                if i == j:
                    continue
                self.nonstable_symbols_used[i] -= self.defined_symbols[j]

    def find_if(self, path, name):
        """
        Find if the symbol is in stablelist or not
        """
        self.all_symbols_used[path].add(name)
        if name in self.matchdata:
            self.stable_symbols[path].add(name)
        else:
            self.nonstable_symbols_used[path].add(name)

    def get_module_name(self, command_line):
        try:
            match = self.HEADER_RE.match(command_line)
            if not match:
                return None
            commands = match.group("cmd").split()

            # Ignore undefined options in parser instead of throwing error
            class IOptParse(OptionParser):
                def error(self, msg):
                    pass

            parser = IOptParse()
            parser.add_option("-k", "--ko")
            opts, _ = parser.parse_args(commands[0:])
            return opts.ko
        except Exception:
            return None


if __name__ == '__main__':
    k = Ksc()
    k.main()
    sys.exit(0)
